From e9faee507047677e6f7523648b95694025ecfa3e Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Thu, 27 Jun 2013 17:26:09 +1000
Subject: [PATCH] Properly Close Inventories

Properly close inventories when unloading and switching worlds.

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 41cace2f..fb699aed 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -940,6 +940,15 @@ public class WorldServer extends World implements GeneratorAccessSeed {
     }
 
     public void unloadChunk(Chunk chunk) {
+        // Spigot Start
+        for (TileEntity tileentity : chunk.getTileEntities().values()) {
+            if (tileentity instanceof IInventory) {
+                for (org.bukkit.entity.HumanEntity h : Lists.newArrayList(((IInventory) tileentity).getViewers())) {
+                    h.closeInventory();
+                }
+            }
+        }
+        // Spigot End
         chunk.C();
     }
 
-- 
2.21.0.windows.1

